### PR TITLE
[Perf] [opengl] Support 'ti.thread_dim(x)' to change invocations per thread for OpenGL

### PR DESCRIPTION
**Pull request by archibate**
Related issue = didn't recall somewhere I promised to @k-ye

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----
This allows user to tweak the stride size to most optimized one.
Feel free to add support in CUDA (@yuanming-hu) and Metal (@k-ye) later :)
I'll add some doc for performance stuffs like `ti.stride_size`, `ti.block_dim` later iapr.

### CODE REVIEW

File: taichi/transforms/offload.cpp [hunk 1]
```diff
@@ -158,6 +159,12 @@ class Offloader {
         offloaded_struct_for->block_dim = for_stmt->block_dim;
       }
     }
+    if (for_stmt->stride_size == 0) {
+      // adaptive
+      offloaded_struct_for->stride_size = 0;
+    } else {
+      offloaded_struct_for->stride_size = for_stmt->stride_size;
+    }
```
Comment (archibate) [inline taichi/transforms/offload.cpp:167]:
```suggestion
    offloaded_struct_for->stride_size = for_stmt->stride_size;
```

**Comment by yuanming-hu** [comment]
Thanks, but I have a question: for grid-stride loops, the stride size equals `gridDim * blockDim`. On CUDA gridDim is fixed and users can either specify `blockDim` or use the default `blockDim`. 

How do we decide `gridDim` (number of workgroups) in OpenGL?

**Comment by archibate** [comment]
> How do we decide gridDim (number of workgroups) in OpenGL?

Specify from C++ code:
```cpp
glDispatchCompute(gridDimX, gridDimY, gridDimZ);
```

> How do we decide blockDim (number of local groups) in OpenGL?

Specify in GLSL at compile-time:
```cpp
layout (local_size_x = blockDimX, ...) in;
```

**Comment by yuanming-hu** [comment]
Thanks for the reply. I understand that in OpenGL you specify the number of workgroups when you dispatch compute, and the size of workgroups in the shader scripts.

I guess the real question here is, how do you determine the value of workgroups? In CUDA one close-to-optimal solution is to always use 20x number of streaming multiprocessor. That's why we don't need a `ti.grid_dim` or `ti.stride_size` option there.

Maybe the reason why we need `ti.stride_size` here is that in OpenGL the number of workgroups is not yet systematically determined by certain rules? 

The other thing is that people usually use `stride_size` to refer to `gridDim * blockDim`. It seems that in the OpenGL backend `stride_size` refers to `gridDim` alone?

**Comment by archibate** [comment]
I feel more confused by the terms..
My understand:

stride size = strides per thread
block dim = threads per block
grid dim = blocks per grid

stride < thread < block < grid

What's your understand?

**Comment by archibate** [comment]
In OpenGL, the default value for `stride per thread` is 1 for maximize performance for normal parallel cases. Non-1 `stride per thread` is only useful when TLS is involved where atomics are performance bottleneck.

**Comment by yuanming-hu** [comment]
I'm following the terms in the original grid-stride loop post: https://developer.nvidia.com/blog/cuda-pro-tip-write-flexible-kernels-grid-stride-loops/

Stride size is defined to be the multiplication of gridDim and blockDim.

**Comment by archibate** [comment]
> I'm following the terms in the original grid-stride loop post: https://developer.nvidia.com/blog/cuda-pro-tip-write-flexible-kernels-grid-stride-loops/
> 
> Stride size is defined to be the multiplication of gridDim and blockDim.

OIC! My idea was:
```py
   ti.stride_size(m)
   for i in range(N):
    ...  // offloaded content
```
```glsl
void main() {  // per-thread
  int _tls_i32_[233];
  for (int i = 0; i < m; i++) {  // per-subthread
    ...  // offloaded content
  }
  atomicAdd(_data_i32_[233], _tls_i32_[233]);
}
```
And seems my term was completely wrong. What should I call the `m` above then?

**Comment by archibate** [comment]
IIUC the method in [nvidia post](https://developer.nvidia.com/blog/cuda-pro-tip-write-flexible-kernels-grid-stride-loops/) is extremely harmful to cache locality:
```
                        memory low -> high
thread id = 0 | 1 | 2 | 3 | 4 | 0 | 1 | 2 | 3 | 4 | 0 | 1 | 2 | 3 | 4
```
While my method is:
```
                      memory low -> high
thread id = 0 | 0 | 0 | 1 | 1 | 1 | 2 | 2 | 2 | 3 | 3 | 3 | 4 | 4 | 4
```

**Comment by archibate** [comment]
I changed the term to `thread_dim`, does this SGTU?

**Comment by archibate** [comment]
```py
for i in range(N):
  do_something()
```
```py
for blockIdx in range(gridDim):  # parallel
  for threadIdx in range(blockDim):   # parallel
    for strideIdx in range(threadDim):  # serial
      do_something()
```

**Comment by archibate** [comment]
I'm merging this in to do more refactor later iapr and since non-GL CI passed and the PR only involve OpenGL for now.
IMO providing `ti.thread_dim` instead of `ti.grid_dim` is optimal for OpenGL since **its `grid_dim` is changeable at run-time**.

**Comment by yuanming-hu** [comment]
Please revert. This is not just an OpenGL change. It changes the IR system shared by all backends. 

Please always wait for at least one approval before merging any PR. I reserve the right to revert PRs that are merged without any approval.

**Pull request merged**